lib/ddtrace/buffer.rb

```diff
@@ -56,5 +70,65 @@ def close
         @closed = true
       end
     end
+
+    # Aggregate metrics:
+    # They reflect buffer activity since last #pop.
+    # These may not be as accurate or as granular, but they
+    # don't use as much network traffic as live stats.
+
+    def measure_accept(trace)
+      @buffer_spans += trace.length
+      @buffer_accepted += 1
+      @buffer_accepted_lengths += trace.length
+    rescue StandardError => e
+      Datadog::Tracer.log.debug("Failed to measure queue accept. Cause: #{e.message} Source: #{e.backtrace.first}")
+    end
+
+    def measure_drop(old_trace, new_trace)
+      @buffer_dropped += 1
+      @buffer_spans -= old_trace.length
+      @buffer_accepted_lengths -= old_trace.length
+      @buffer_accepted += 1
+      @buffer_spans += new_trace.length
+      @buffer_accepted_lengths += new_trace.length
+    rescue StandardError => e
+      Datadog::Tracer.log.debug("Failed to measure queue drop. Cause: #{e.message} Source: #{e.backtrace.first}")
+    end
+
+    def measure_pop(traces)
+      # Accepted
+      Debug::Health.metrics.queue_accepted(@buffer_accepted)
+      Debug::Health.metrics.queue_accepted_lengths(@buffer_accepted_lengths)
+      Debug::Health.metrics.queue_accepted_size { measure_traces_size(traces) }
+
+      # Dropped
+      Debug::Health.metrics.queue_dropped(@buffer_dropped)
+
+      # Queue gauges
+      Debug::Health.metrics.queue_max_length(@max_size)
+      Debug::Health.metrics.queue_spans(@buffer_spans)
+      Debug::Health.metrics.queue_length(traces.length)
+      Debug::Health.metrics.queue_size { measure_traces_size(traces) }
```

Comment:
I think measuring in this way concerns me a little.

It probably doesn't end up being too much of a computational burden right now since the max buffer size is only 100 traces, but when we increase it to 1000, it'll have way more of an impact.

Reply:
Here's a benchmark for measuring traces with two spans each:

```
Rehearsal --------------------------------------------------
100 traces:      0.000000   0.000000   0.000000 (  0.002725)
1000 traces:     0.030000   0.000000   0.030000 (  0.025894)
10000 traces:    0.230000   0.070000   0.300000 (  0.304395)
100000 traces:   2.250000   0.730000   2.980000 (  2.976566)
----------------------------------------- total: 3.310000sec

                     user     system      total        real
100 traces:      0.000000   0.000000   0.000000 (  0.000947)
1000 traces:     0.010000   0.000000   0.010000 (  0.007790)
10000 traces:    0.080000   0.000000   0.080000 (  0.079351)
100000 traces:   0.770000   0.000000   0.770000 (  0.772285)
```

So for 1000 traces (or 2000 spans) it'd take something like ~10ms. It's very likely there will be more than 2 spans per trace, although I don't know how many on average. Conceivably, 1000 traces with 200 spans each I would expect to take at least 770ms. That would be in addition to the time it takes to encode and flush the traces. I might agree with you that for a diagnostic, this is too expensive.

I don't think there's a cheaper way of calculating the size while still preserving any usefulness to this metric. I would suggest scrapping it.